### PR TITLE
add typings for message.py

### PR DESCRIPTION
**Pull request by Zomatree**
## Summary

Adds typings to everything inside message.py.

## Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: discord/message.py [hunk 1]
```diff
@@ -988,18 +1012,18 @@ async def delete(self, *, delay=None):
...
                 except HTTPException:
                     pass
 
-            asyncio.create_task(delete())
+            asyncio.create_task(delete(delay))
         else:
             await self._state.http.delete_message(self.channel.id, self.id)
 
-    async def edit(self, **fields):
+    async def edit(self, **fields) -> None:
```
Comment (bijij) [inline discord/message.py:1026]: fields  should probably be `Any`

File: discord/types/reaction.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+from typing import TypedDict
+
+from .emoji import PartialEmoji
+
+
+class Reaction(TypedDict):
+    count: int
+    me: bool
+    emoji: PartialEmoji
```
Comment (NCPlayz) [inline discord/types/reaction.py:35]: this is already in types/message.py.

Comment (Rapptz) [reply discord/types/reaction.py:35]: Yeah this entire file is unnecessary.

File: discord/message.py [hunk 3]
```diff
@@ -552,7 +574,9 @@ class Message(Hashable):
                  '_cs_system_content', '_cs_guild', '_state', 'reactions', 'reference',
                  'application', 'activity', 'stickers')
 
-    def __init__(self, *, state, channel, data):
+    author: Union[Member, User]
```
Comment (Rapptz) [inline discord/message.py:577]: Why is this here?

File: discord/message.py [hunk 4]
```diff
@@ -460,17 +481,17 @@ class Message(Hashable):
...
     content: :class:`str`
         The actual contents of the message.
-    nonce
+    nonce: Union[:class:`str`, :class:`int`]
         The value used by the discord guild and the client to verify that the message is successfully sent.
         This is not stored long term within Discord's servers and is only used ephemerally.
     embeds: List[:class:`Embed`]
         A list of embeds the message has.
-    channel: Union[:class:`abc.Messageable`]
+    channel: :class:`abc.Messageable`
```
Comment (Rapptz) [inline discord/message.py:494]:
```suggestion
    channel: Union[:class:`TextChannel`, :class:`DMChannel`, :class:`GroupChannel`]
```

File: discord/message.py [hunk 5]
```diff
@@ -628,7 +649,7 @@ def _add_reaction(self, data, emoji, user_id):
 
         return reaction
 
-    def _remove_reaction(self, data, emoji, user_id):
+    def _remove_reaction(self, data: ReactionPayload, emoji: Union[Emoji, PartialEmoji, str], user_id: int) -> Reaction:
```
Comment (Rapptz) [inline discord/message.py:652]: I think `Union[Emoji, PartialEmoji, str]` should probably be aliased since it's used throughout the file. Something like `EmojiInputType` or something within the `TYPE_CHECKING` block.

File: discord/message.py [hunk 6]
```diff
@@ -664,7 +685,7 @@ def _update(self, data):
         # otherwise they overwrite each other which is undesirable.
         # Since there's no good way to do this we have to iterate over every
         # handler rather than iterating over the keys which is a little slower
-        for key, handler in self._HANDLERS:
+        for key, handler in self._HANDLERS:  # type: ignore
```
Comment (Rapptz) [inline discord/message.py:688]: To fix this type ignore you need to add

```py
_HANDLERS: ClassVar[List[Tuple[str, Callable[..., None]]]]
```

File: discord/message.py [hunk 7]
```diff
@@ -673,56 +694,56 @@ def _update(self, data):
                 handler(self, value)
 
         # clear the cached properties
-        for attr in self._CACHED_SLOTS:
+        for attr in self._CACHED_SLOTS:  # type: ignore
```
Comment (Rapptz) [inline discord/message.py:697]: To fix this type ignore you need to add

```py
_CACHED_SLOTS: ClassVar[List[str]]
```

File: discord/message.py [hunk 8]
```diff
@@ -552,7 +573,7 @@ class Message(Hashable):
                  '_cs_system_content', '_cs_guild', '_state', 'reactions', 'reference',
                  'application', 'activity', 'stickers')
 
-    def __init__(self, *, state, channel, data):
+    def __init__(self, *, state: ConnectionState, channel: Messageable, data: MessagePayload):
```
Comment (Rapptz) [inline discord/message.py:576]:
```suggestion
    def __init__(self, *, state: ConnectionState, channel: Union[TextChannel, DMChannel, GroupChannel], data: MessagePayload):
```

File: discord/message.py [hunk 9]
```diff
@@ -754,29 +775,29 @@ def _handle_mentions(self, mentions):
...
+    def _handle_mention_roles(self, role_mentions: List[int]) -> None:
         self.role_mentions = []
         if isinstance(self.guild, Guild):
             for role_id in map(int, role_mentions):
                 role = self.guild.get_role(role_id)
                 if role is not None:
                     self.role_mentions.append(role)
 
-    def _rebind_channel_reference(self, new_channel):
+    def _rebind_channel_reference(self, new_channel: Messageable) -> None:
```
Comment (Rapptz) [inline discord/message.py:786]:
```suggestion
    def _rebind_channel_reference(self, new_channel: Union[TextChannel, DMChannel, GroupChannel]) -> None:
```

File: discord/message.py [hunk 10]
```diff
@@ -1294,7 +1315,7 @@ async def clear_reactions(self):
         """
         await self._state.http.clear_reactions(self.channel.id, self.id)
 
-    async def reply(self, content=None, **kwargs):
+    async def reply(self, content: str = None, **kwargs) -> Message:
```
Comment (Rapptz) [inline discord/message.py:1318]:
```suggestion
    async def reply(self, content: Optional[str] = None, **kwargs) -> Message:
```

**Review by Rapptz — CHANGES_REQUESTED** [review]

**Review by Rapptz — APPROVED** [review]
